File: 201/Graph/Connected-Components/Number-Of-Enclaves/solution.cs
public class Solution {
    private int[,] dirns = new int[,] {{-1,0}, {1,0}, {0,1}, {0, -1}};
    public int NumEnclaves(int[][] grid) {
        int rowSz = grid.Length, colSz = grid[0].Length, result = 0;
        bool[,] visited = new bool[rowSz, colSz];
        for(int c = 0; c < colSz; ++c) {
            //First Row
            Visit(grid, 0, c, visited);

            //Last Row
            Visit(grid, rowSz-1, c, visited);
        }

        for(int r = 0; r < rowSz; ++r) {
            //First COl
            Visit(grid, r, 0, visited);

            //Last Col
            Visit(grid, r, colSz-1, visited);
        }

        for(int r = 0; r < rowSz; ++r) {
            for (int c = 0; c < colSz; ++c) {
                if(grid[r][c] == 1 && !visited[r,c])
                    ++result;
            }
        }
        return result;
    }

    private void Visit(int[][] grid, int row, int col, bool[,] visited) {
        if(grid[row][col] == 0 || visited[row,col])
            return;
        visited[row,col] = true;
        int rowSz = grid.Length, colSz = grid[0].Length;
        for(int d = 0; d < 4; ++d) {
            int nr = row +dirns[d,0];
            int nc = col + dirns[d,1];
            if(nr < 0 || nc < 0 || nr >= rowSz || nc >= colSz || visited[nr,nc] || grid[nr][nc] == 0)
                continue;
            Visit(grid, nr,nc, visited);
            
        }
    }
}

File: 201/Graph/Connected-Components/Number-Of-Enclaves/solution.py
# https://leetcode.com/problems/number-of-enclaves/description/
class Solution(object):
    def numEnclaves(self, grid):
        """
        :type grid: List[List[int]]
        :rtype: int
        """
        
        rowSz,colSz = len(grid), len(grid[0])
        visited = [[False for _ in range(colSz)] for _ in range(rowSz)]
        dirns = [
            [-1,0],
            [1,0],
            [0,1],
            [0,-1]
        ]

        def dfs(r,c):
            # visite (r,c)
            visited[r][c] = True

            # explore neighbours
            for dirn in dirns:
                nr = r + dirn[0]
                nc = c + dirn[1]
                if nr < 0 or nr >= rowSz or nc < 0 or nc >= colSz:
                    continue
                if visited[nr][nc] or grid[nr][nc] == 0:
                    continue
                dfs(nr,nc)
                
        
        # first column and last col
        for r in range(rowSz):
            if grid[r][0] == 1:
                dfs(r,0)
            
            if grid[r][colSz-1] == 1:
                dfs(r,colSz-1)


        # first row and last row
        for c in range(colSz):
            if grid[0][c] == 1:
                dfs(0,c)

            if grid[rowSz-1][c] == 1:
                dfs(rowSz-1,c)

        count = 0
        for r in range(rowSz):
            for c in range(colSz):
                if grid[r][c] == 1 and visited[r][c]==False:
                    count+=1
        return count